blobby/blob.py

```diff
@@ -1,248 +0,0 @@
-    pass
-
-
-def inherit_docstrings(cls: Type[T]) -> Type[T]:
-    for m in getmembers(cls):
-        if isfunction(m) and not m.__doc__:
-            m.__doc__ = getdoc(m)  # pragma: no cover
-
-    return cls
-
-
-@inherit_docstrings
-class BlobReader(io.RawIOBase):
-    """A read-only, file-like object wrapping an S3 blob."""
-    WHENCE_LIST = ', '.join(
-        str(t) for t in (io.SEEK_SET, io.SEEK_CUR, io.SEEK_END)
-    )
-
-    def __init__(self, s3_bucket: S3Bucket, key: str) -> None:
-        self.object = s3_bucket.Object(key=key)
-        self.position = 0
-
-    @contextmanager
-    def as_fifo(self) -> Iterator[BinaryIO]:
-        """Context manager which creates a FIFO and writes the blob to it.
-
-        Returns: an open file object of the read end of the FIFO, which
-            will be closed by the context manager.
-        """
-        with tempfile.TemporaryDirectory() as dir:
-            path = os.path.join(dir, 'fifo')
-            os.mkfifo(path, mode=0o600)
-            writer = threading.Thread(
-                target=self._fifo_writer, args=(path,)
-            )
-            writer.start()
-            with open(path, 'rb') as fd_r:
-                yield fd_r
-                writer.join()
-
-    def _fifo_writer(self, path: str) -> None:
-        self.seek(0)
-        with open(path, 'wb') as fd_w:
-            fd_w.write(self.read())
-
-    @contextmanager
-    def as_tempfile(self, chunk_size: int = 1000000) -> Iterator[IO[bytes]]:
-        """Context manager which synchronously downloads the object to
-        a temporary file and returns a file descriptor for it.
-
-        Returns: the open temporary file, which will be closed by the
-            context manager.
-        """
-        with tempfile.TemporaryFile('w+b') as fd:
-            chunk = self.read(chunk_size)
-            while chunk:
-                fd.write(chunk)
-                chunk = self.read(chunk_size)
-            fd.seek(0)
-            yield fd
-
-    def seek(self, offset: int, whence: int = io.SEEK_SET) -> int:
-        old_position = self.position
-
-        if whence == io.SEEK_SET:
-            self.position = offset
-        elif whence == io.SEEK_CUR:
-            self.position += offset
-        elif whence == io.SEEK_END:
-            self.position = self.object.content_length + offset
-        else:
-            raise ValueError(
-                f'whence must be one of {self.WHENCE_LIST}, not {whence}.'
-            )
-        if self.position < 0:
-            self.position = old_position
-            raise ValueError(
-                f'seek would result in negative position {self.position}'
-            )
-
-        return self.position
-
-    def seekable(self) -> bool:
-        return True
-
-    def read(self, size: int = -1) -> bytes:
-        # GetObject will be unhappy if we try to read from beyond the
-        # end of the object.
-        if self.position >= len(self):
-            return b''
```

Comment:
When might this happen? Does this indicate an error somewhere e.g. surely `self.position` should not reach past the end of `self`? If this is indeed an error case then raising an explicit exception is probably a good idea.

Reply:
This is to match the API; if you open an empty file and `seek(100); read()` on it you'll just get an empty string (or bytes if in binary mode), not an error. It doesn't really make sense to me either but presumably that in turn is modelled on the underlying C/POSIX specs...